li-apache-kafka-clients/src/main/java/com/linkedin/kafka/clients/producer/LiKafkaFederatedProducerImpl.java

```diff
@@ -355,7 +361,68 @@ public void close() {
 
   @Override
   public void close(long timeout, TimeUnit timeUnit) {
-    throw new UnsupportedOperationException("Not implemented yet");
+    if (_producers.isEmpty()) {
+      LOG.warn("No producers to close for cluster group {}", _clusterGroup);
+      return;
+    }
+
+    LOG.info("Closing LiKafkaProducer for cluster group {} in {} {}...", _clusterGroup, timeout, timeUnit);
+
+    long startTimeMs = System.currentTimeMillis();
+    long budgetMs = timeUnit.toMillis(timeout);
+    long deadlineTimeMs = startTimeMs + budgetMs;
+
+    List<Thread> taskThreads = new ArrayList<>();
+    for (Map.Entry<ClusterDescriptor, PerClusterProducerContext<K, V>> entry : _producers.entrySet()) {
+      Thread taskThread = new Thread(new CloseTask(entry.getValue(), deadlineTimeMs));
+      taskThreads.add(taskThread);
+      taskThread.start();
+    }
+
+    for (Thread taskThread : taskThreads) {
+      try {
+        taskThread.join();
+      } catch (InterruptedException e) {
+        throw new KafkaException("Fail to close all producers for cluster group " + _clusterGroup, e);
+      }
+    }
+
+    LOG.info("LiKafkaProducer close for cluster group {} complete in {} milliseconds", _clusterGroup,
+        (System.currentTimeMillis() - startTimeMs));
+  }
+
+  private class CloseTask implements Runnable {
+    private PerClusterProducerContext<K, V> _context;
+    private long _deadlineTimeMs;
+
+    CloseTask(PerClusterProducerContext<K, V> context, long deadlineTimeMs) {
+      _context = context;
+      _deadlineTimeMs = _deadlineTimeMs;
+    }
+
+    public void run() {
+      LOG.info("Closing LiKafkaProducer for cluster {}", _context.cluster());
+      LongAdder numThreadsInSend = _context.numThreadsInSend();
```

Comment:
I want to understand the association between the number of producers per cluster (I am assuming cluster here is actually a cluster group) and the number of threads in Sends. I am not clear on why we need a `numThreadsInSend` when there is a similar logic in `LiKafkaProducerImpl` as well.

Reply:
The "cluster" here is actually a physical cluster, not a cluster group, and we maintain a separate numThreadsInSend per physical cluster via context.

I copied the send() implementation from LiKafkaProducerImpl here. I am not calling LiKafkaProducerImpl#send().

Reply:
Per our discussion, I restructured the code to directly create and use per-cluster LiKafkaProducer.